currencies/rate_converter_test.go

```diff
@@ -557,25 +271,39 @@ func TestRatesAreNeverStale(t *testing.T) {
 		},
 	}
 
+	initialFakeTime := time.Date(2018, time.September, 12, 30, 0, 0, 0, time.UTC)
+	fakeTime := &FakeTime{time: initialFakeTime}
+
 	// Execute:
-	currencyConverter := currencies.NewRateConverter(
+	currencyConverter := currencies.NewConfiguredRateConverter(
 		&http.Client{},
 		mockedHttpServer.URL,
-		time.Duration(100)*time.Millisecond,
-		time.Duration(0)*time.Millisecond,
+		100*time.Millisecond,
+		0*time.Millisecond, // stale rates threshold
+		fakeTime,
 	)
 
-	// Verify:
-	// Rates are valid at t=0 and are then invalid at 100ms
-	assert.Equal(t, currencyConverter.Rates(), expectedRates, "Rates() should return expected rates")
+	// First Update call is successful and yields valid rates
+	err1 := currencyConverter.Run()
+	assert.Nil(t, err1)
+
+	// Verify rates are valid and last update timestamp is correct
+	assert.Equal(t, expectedRates, currencyConverter.Rates(), "Conversions.Rates weren't the expected ones")
+	assert.Equal(t, fakeTime.time, currencyConverter.LastUpdated(), "LastUpdated should be set")
+
+	// Advance time so the current time is well past the the time the rates were last updated
+	fakeTime.time = initialFakeTime.Add(24 * time.Hour)
 
-	time.Sleep(500 * time.Millisecond)
-	// Rates have been invalid for ~400ms
-	assert.Equal(t, currencyConverter.Rates(), expectedRates, "Rates() should return expected rates")
+	// Second Update call results in error but rates from a day ago are still valid
+	err2 := currencyConverter.Run()
+	assert.NotNil(t, err2)
+
+	// Verify rates are valid and last update ts is correct
+	assert.Equal(t, expectedRates, currencyConverter.Rates(), "Conversions.Rates weren't the expected ones")
+	assert.Equal(t, initialFakeTime, currencyConverter.LastUpdated(), "LastUpdated should be set")
 }
 
 func TestRace(t *testing.T) {
```

Comment:
I guess we can leave the logic alone. I don't follow it's design. 50 clients seems like a lot and the random intervals only decrease the change of a race condition and lower the effectiveness of the test. Since order of operations will be impossible to predict here, this looks like a stub for Go's memory race detector to report on unprotected access.

Reply:
Yeah I don't understand it either. I took a look at the issue that resulted in this test but I still hesitated to edit this because I was afraid I would be removing coverage of some edge case I'm not seeing. I'll take a stab at writing a simpler race test given it seems overcomplicated to both of us.